hir/src/builder.rs

```diff
@@ -652,7 +652,8 @@ pub trait InstBuilder<'f>: InstBuilderBase<'f> {
             addr,
             ty: ty.clone(),
         });
-        into_first_result!(self.build(data, ty, span))
+        let addr_ty = self.data_flow_graph().value_type(addr).clone();
```

Comment:
@bitwalker I'm passing "full" `Type::Ptr` here because it's expected in `Opcode::results()`.

Reply:
We should probably just fix `results` actually, I don't think there's much benefit in special casing the loads with regard to the controlling type in any case.

As an aside, the way loads are handled is going to change, to distinguish loads of "native" pointers (in a word-addressable address space), and aligned vs unaligned loads of pointers in the default byte-addressable address space. Aligned vs unaligned here refers to whether the address being loaded is aligned to a word or field element boundary (32-byte or 8-byte alignment respectively), in addition to whether the load can be guaranteed to not cross a word boundary. If those properties hold, one can use the aligned loads, which are much more efficient. If not, unaligned loads must be used, which perform extra checks and handle the case of loading across a word boundary.

Those changes fall out from the fact that Miden VMs address space is word-addressable, i.e. each unique address refers to an entire word (4 field elements). There are two ops in MASM which allow addressing the first field element of each word, and I'm working through a proposal to extend those to allow addressing any field element of the word, which would make the Miden VM address space both word- and felt-addressable. In either case though, we have to map our byte-addressable IR to this paradigm, and since we only have the ability to load/store words or field elements, we have to handle cases where a value spans one or both of those. Aligned loads can be made much more efficient because if we know an address is word-aligned, and the type is <= the size of a word (or felt-aligned, <= the size of a felt), we can fetch the value with a single instruction. Unaligned loads require us to figure out whether we can load data from a pair of felts in a single word, or a pair of words, and then shift the values so that they are aligned on the operand stack.

As an additional upcoming restriction, the maximum size of a value that I think we'll allow loads/stores for is <= a single word. We could support more than that, but we can't access more than 16 elements on the operand stack, and using more than a word for temporaries leaves little room for other live values. I would limit it further to single field elements, but there are a number of instructions in Miden that operate on words, so we will necessarily need to permit loading/storing words for use with those instructions. It remains to be seen how common it is for there to be more than 16 slots of the operand stack in use at once.

Reply:
Makes sense. Thanks for the detailed explanation! I saw your proposal earlier today in miden-vm repo.